src/Tests/InstrumentationEngineLibTests/CMakeLists.txt

```diff
@@ -58,14 +65,21 @@ target_link_libraries(
     Common.Lib
     palrt
     corguids
-    :libuuid.so
-    :libunwind-x86_64.so
-    stdc++fs
     gtest_main
 )
 
+if (CMAKE_SYSTEM_NAME STREQUAL Linux)
+    target_link_libraries(${PROJECT_NAME} stdc++fs)
+elseif(CMAKE_SYSTEM_NAME STREQUAL Darwin)
+    target_link_libraries(${PROJECT_NAME}
+        iconv
```

Comment:
>iconv [](http://example.com/codeflow?start=8&length=5)

is this going to cause static linkage to libiconv? What is the license of libiconv on a mac?

Reply:
Yes, the linker uses a shared library unless only a static one is available or a `-static` option is given. I don't think it makes sense to try to force dynamic linking somehow. Many systems won't even have a static library; it would take a very unusual configuration to only have a static one. This is especially true of libdl (in fact I don't think I've ever seen a libdl.a).

A CI machine would have shared objects, so it'll get linked dynamically and the binaries can be freely redistributed. If a user has a weird configuration with only static libraries, it's their decision and responsibility not to redistribute the statically linked binary if it's against the license (which I believe is not 100% clear).

BTW, this iconv is for test code which isn't redistributed in a compiled form anyway.

Reply:
IANAL, but if my understanding is correct, statically linking to an LGPL binary implies that the distribution of the new dependent binary must be redistributed as LGPL... not so if it is statically linked. There is, at least, some controversy around it. We would like to be able to release all of our binaries as MIT and avoid controversy. We would also like to redistribute them under alternate licenses (and allow other developers to redistribute under alternate licenses). In any case, the code remains MIT.

Also, it isn't entirely clear whether or not uploading the resulting binaries as artifacts of a CI counts as "distribution"... so there is that.

All-in-all, I think that we would like to make dynamic linkage as explicit as possible so that we don't accidentally make a mistake in the future.

Reply:
IANAL either ;)

FWIW, there doesn't seem to be a way to explicitly instruct ld64 (OSX linker) to only search for dynamic libraries. I tried using `-lazy-l` but it appears to be deprecated and falls back to normal link. I verified that on it indeed links dynamically to libiconv.dylib. Libdl appears to be a compatibility layer and it's not linked directly, instead the calls are replaced by _dl* system builtins (imported dynamically too, probably by the dyld interpreter on execution); it is kind of a special case library, because it really is an interface to the dynamic linker which loads and executes the binary image.